### ISSUE TITLE
Refactor stochastic optimizers

### ISSUE THREAD

**Issue by accosmin** [accosmin/nano#137]
Simplify & clarify the implementation of the stochastic optimizers:
* have the parameter smoothing optional (see loop.h)
* explicit stochastic & batch problems passed
* always update the batch values (value, gradient) after each epoch
* thus can check for convergence after each epoch using a given epsilon
* extend the learning rate schedule to: fixed, exponential decay, t-inverse (see the SVRG paper - experimentation section)

**Comment by accosmin** [comment]
To also fix:
* the training loss/error is computed twice for the batch optimizers (see batch_params_t::ulog)

**Comment by accosmin** [comment]
Extend test_affine to support other criteria

**Comment by accosmin** [comment]
Also use a single accumulator for training models: each one (lacc/gacc) will create a copy of the models and parameters. This may be problematic for large models.
Also check if it is possible to merge accumulator_t and criterion_t.

**Comment by accosmin** [comment]
Remove stoch_params_t::config_t and trainer_result's config_t: this can be achieved easily with nano::to_params

**Comment by accosmin** [comment]
SG/SGM should never fail during benchmark_stoch for convex functions.
